### PR TITLE
Port test_web.py to Python 3

### PR DESCRIPTION
**Pull request by itamarst**
Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3596

### CODE REVIEW

File: src/allmydata/util/jsonbytes.py [hunk 1]
```diff
@@ -13,34 +13,41 @@
 if PY2:
     from future.builtins import filter, map, zip, ascii, chr, hex, input, next, oct, open, pow, round, super, bytes, dict, list, object, range, str, max, min  # noqa: F401
 
-
 import json
 
 
+def _bytes_to_unicode(obj):
+    """Convert keys of dicts from bytes to unicode, recursively."""
```
Comment (chadwhitacre) [inline src/allmydata/util/jsonbytes.py:20]: Converts values as well, afaict.

File: src/allmydata/webish.py [hunk 2]
```diff
@@ -44,6 +44,42 @@
     StoragePlugins,
 )
 
+
+if PY2:
+    FileUploadFieldStorage = FieldStorage
+else:
+    class FileUploadFieldStorage(FieldStorage):
+        """
+        Do terrible things to ensure files are still bytes.
```
Comment (chadwhitacre) [inline src/allmydata/webish.py:53]: [Looks like](https://app.circleci.com/pipelines/github/tahoe-lafs/tahoe-lafs/1864/workflows/2677d95a-d206-4a4b-af5c-6edfb10dcd71/jobs/44212) CI agrees with you that they're terrible:

```
typechecks create: /root/project/.tox/typechecks
typechecks installdeps: mypy, git+https://github.com/Shoobx/mypy-zope, git+https://github.com/warner/foolscap
typechecks installed: attrs==20.3.0,Automat==20.2.0,cffi==1.14.4,constantly==15.1.0,cryptography==3.3.1,foolscap @ git+https://github.com/warner/foolscap@291df027d5770f5edd6df01368de3e2cf273ccdf,hyperlink==21.0.0,idna==3.1,incremental==17.5.0,mypy==0.800,mypy-extensions==0.4.3,mypy-zope @ git+https://github.com/Shoobx/mypy-zope@ca0fe617f0303e60faf944e02b47a6b2717c3921,pyasn1==0.4.8,pyasn1-modules==0.2.8,pycparser==2.20,PyHamcrest==2.0.2,pyOpenSSL==20.0.1,service-identity==18.1.0,six==1.15.0,Twisted==20.3.0,typed-ast==1.4.2,typing-extensions==3.7.4.3,zope.event==4.5.0,zope.interface==5.2.0,zope.schema==6.0.1
typechecks run-test-pre: PYTHONHASHSEED='787671381'
typechecks run-test: commands[0] | mypy src
src/allmydata/webish.py:68: error: Signature of "filename" incompatible with supertype "FieldStorage"
Found 1 error in 1 file (checked 284 source files)
ERROR: InvocationError for command /root/project/.tox/typechecks/bin/mypy src (exited with code 1)
___________________________________ summary ____________________________________
ERROR:   typechecks: commands failed
```

File: src/allmydata/util/_python3.py [hunk 3]
```diff
@@ -196,5 +196,6 @@
     "allmydata.test.web.test_root",
     "allmydata.test.web.test_status",
     "allmydata.test.web.test_util",
+    "allmydata.test.web.test_web",
```
Comment (chadwhitacre) [inline src/allmydata/util/_python3.py:199]: Big diff for one line. :o)

File: src/allmydata/test/common_util.py [hunk 4]
```diff
@@ -203,6 +203,11 @@ def flip_one_bit(s, offset=0, size=None):
 class ReallyEqualMixin(object):
     def failUnlessReallyEqual(self, a, b, msg=None):
         self.assertEqual(a, b, msg)
+        # Make sure unicode strings are a consistent type:
```
Comment (chadwhitacre) [inline src/allmydata/test/common_util.py:206]: If they're not a consistent type then how would we get here, since `"foo" != b"foo"`? Doesn't the `assertEqual` just before fail in that case?

File: src/allmydata/test/test_util.py [hunk 5]
```diff
@@ -493,10 +493,10 @@ class JSONBytes(unittest.TestCase):
     def test_encode_bytes(self):
         """BytesJSONEncoder can encode bytes."""
         data = {
-            b"hello": [1, b"cd"],
+            b"hello": [1, b"cd", {b"abc": 123}],
         }
         expected = {
-            u"hello": [1, u"cd"],
+            u"hello": [1, u"cd", {u"abc": 123}],
```
Comment (chadwhitacre) [inline src/allmydata/test/test_util.py:499]: Kinda light on the coverage since e.g. we decode values as well. Consider expanding.

File: src/allmydata/web/root.py [hunk 6]
```diff
@@ -293,7 +294,7 @@ def _describe_known_servers(self, broker):
             self._describe_server(server)
             for server
             in broker.get_known_servers()
-        ))
+        ), key=lambda o: sorted(o.items()))
```
Comment (chadwhitacre) [inline src/allmydata/web/root.py:297]: Why is this necessary?

**Review by chadwhitacre — CHANGES_REQUESTED** [review]
Mostly the usual b"whack-a-mole" ... and plenty of it! The only blocker I see is the mypy complaint. Good luck with that. 🖖

File: src/allmydata/util/_python3.py (hunk 3 above)
Comment (itamarst) [reply src/allmydata/util/_python3.py:199]: 270 tests!

File: src/allmydata/test/common_util.py (hunk 4 above)
Comment (itamarst) [reply src/allmydata/test/common_util.py:206]: It's because it's the Future newstr (backported Unicode type) vs. Python 2 native unicode type. So they're equal, and _logically_ the same type, but have different types in practice. I'll update the comment.

File: src/allmydata/web/root.py (hunk 6 above)
Comment (itamarst) [reply src/allmydata/web/root.py:297]: Dicts don't have order in Python 3.

```
>>> {} > {}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '>' not supported between instances of 'dict' and 'dict'
```

File: src/allmydata/webish.py (hunk 2 above)
Comment (itamarst) [reply src/allmydata/webish.py:53]: I just told mypy to not type check that function :(

**Comment by itamarst** [comment]
Thanks for the review! I have addressed all comments.

**Review by chadwhitacre — APPROVED** [review]
lgtm, though now codecov is failing, and my attempts to re-run seem to be failing silently. 😞

File: src/allmydata/web/directory.py [hunk 7]
```diff
@@ -1290,7 +1298,7 @@ def render_HTML(self, req):
                      + stats.get("size-mutable-files", 0)
                      + stats.get("size-directories", 0))
             output += "size: %d\n" % total
-        return output
+        return output.encode("utf-8")
```
Comment (exarkun) [inline src/allmydata/web/directory.py:1301]: Maybe this isn't necessary given the change to `MultiFormatResource` (not that it really seems to hurt, either).

File: src/allmydata/test/web/test_web.py [hunk 8]
```diff
@@ -494,13 +505,13 @@ def failUnlessIsFooJSON(self, res):
...
-        kidnames = sorted([unicode(n) for n in data[1]["children"]])
+        kidnames = sorted([str(n) for n in data[1]["children"]])
         self.failUnlessEqual(kidnames,
                              [self._htmlname_unicode, u"bar.txt", u"baz.txt",
                               u"blockingfile", u"empty", u"n\u00fc.txt", u"quux.txt", u"sub"])
-        kids = dict( [(unicode(name),value)
+        kids = dict( [(str(name),value)
                       for (name,value)
-                      in data[1]["children"].iteritems()] )
+                      in list(data[1]["children"].items())] )
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:514]: The `list` seems superfluous but also harmless

File: src/allmydata/test/web/test_web.py [hunk 9]
```diff
@@ -1721,26 +1741,26 @@ def test_PUT_NEWFILEURL_unlinked_mdmf(self):
...
+        d.addCallback(lambda json: self.failUnlessIn(b"SDMF", json))
         return d
 
     @inlineCallbacks
     def test_PUT_NEWFILEURL_unlinked_bad_format(self):
         contents = self.NEWFILE_CONTENTS * 300000
         yield self.assertHTTPError(self.webish_url + "/uri?format=foo", 400,
-                                   "Unknown format: foo",
+                                   "Unknown format:",
                                    method="put", data=contents)
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:1764]: What happened to the "foo"?  Did it turn in to `b"foo"` or something?

I guess it would be nice to continue to make some kind of assertion about "foo" in this string... although even more interesting would be a test for `format=<some csrf string>` which demonstrates it gets quoted correctly (but clearly this is beyond the scope of just porting).

File: src/allmydata/test/web/test_web.py [hunk 10]
```diff
@@ -1793,9 +1813,9 @@ def test_PUT_NEWFILEURL_replace(self):
 
     def test_PUT_NEWFILEURL_bad_t(self):
         d = self.shouldFail2(error.Error, "PUT_bad_t", "400 Bad Request",
-                             "PUT to a file: bad t=bogus",
+                             "PUT to a file: bad t=",
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:1816]: Same comment about "bogus" here as "foo" above

File: src/allmydata/test/web/test_web.py [hunk 11]
```diff
@@ -1928,7 +1948,7 @@ def test_GET_FILEURL_badtype(self):
     def test_CSS_FILE(self):
         d = self.GET("/tahoe.css", followRedirect=True)
         def _check(res):
-            CSS_STYLE=re.compile('toolbar\s{.+text-align:\scenter.+toolbar-item.+display:\sinline',re.DOTALL)
+            CSS_STYLE=re.compile(b'toolbar\s{.+text-align:\scenter.+toolbar-item.+display:\sinline',re.DOTALL)
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:1951]: :bird: :notes: :notes: :notes:

File: src/allmydata/test/web/test_web.py [hunk 12]
```diff
@@ -2206,11 +2226,14 @@ def test_POST_DIRURL_deepsize(self):
         def _got_json(data):
             self.failUnlessReallyEqual(data["finished"], True)
             size = data["size"]
-            self.failUnless(size > 1000)
+            # Match calculation of text value size below:
+            self.failUnless(
+                size.get("size-directories", 0) + size.get("size-mutable-files", 0) +
+                size.get("size-immutable-files", 0) > 1000)
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:2232]: was this a non-deterministic test?  or does it share state with some other test?  or is this just a free enhancement? :)

File: src/allmydata/test/web/test_web.py [hunk 13]
```diff
@@ -2321,7 +2344,7 @@ def test_PUT_NEWDIRURL_sdmf(self):
     def test_PUT_NEWDIRURL_bad_format(self):
         url = (self.webish_url + self.public_url +
                "/foo/newdir=?t=mkdir&format=foo")
-        yield self.assertHTTPError(url, 400, "Unknown format: foo",
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:2324]: Another "t=foo"

File: src/allmydata/test/web/test_web.py [hunk 14]
```diff
@@ -2354,7 +2377,7 @@ def test_POST_NEWDIRURL_sdmf(self):
     def test_POST_NEWDIRURL_bad_format(self):
         url = (self.webish_url + self.public_url +
                "/foo/newdir?t=mkdir&format=foo")
-        yield self.assertHTTPError(url, 400, "Unknown format: foo",
+        yield self.assertHTTPError(url, 400, "Unknown format:",
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:2380]: Another "t=foo" (sigh, why so many)

File: src/allmydata/test/web/test_web.py [hunk 15]
```diff
@@ -2431,16 +2454,16 @@ def test_POST_NEWDIRURL_initial_children_bad_format(self):
         (newkids, caps) = self._create_initial_children()
         url = (self.webish_url + self.public_url +
                "/foo/newdir?t=mkdir-with-children&format=foo")
-        yield self.assertHTTPError(url, 400, "Unknown format: foo",
-                                   method="post", data=json.dumps(newkids))
+        yield self.assertHTTPError(url, 400, "Unknown format:",
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:2457]: Another! "t=foo"!

File: src/allmydata/test/web/test_web.py [hunk 16]
```diff
@@ -2555,7 +2578,7 @@ def _got_subdir(subdir):
     def test_PUT_NEWDIRURL_mkdirs_bad_format(self):
         url = (self.webish_url + self.public_url +
                "/foo/subdir/newdir?t=mkdir&format=foo")
-        yield self.assertHTTPError(url, 400, "Unknown format: foo",
+        yield self.assertHTTPError(url, 400, "Unknown format:",
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:2581]: Ugg, another "t=foo"

File: src/allmydata/test/web/test_web.py [hunk 17]
```diff
@@ -2820,18 +2846,18 @@ def _got_json(raw):
...
+        d.addCallback(_check_upload_unlinked, "mdmf", b"URI:MDMF")
         return d
 
     @inlineCallbacks
     def test_POST_upload_bad_format_unlinked(self):
         url = self.webish_url + "/uri?t=upload&format=foo"
         body, headers = self.build_form(file=("foo.txt", self.NEWFILE_CONTENTS * 300000))
         yield self.assertHTTPError(url, 400,
-                                   "Unknown format: foo",
+                                   "Unknown format:",
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:2860]: another ... "t=foo" ...

File: src/allmydata/test/web/test_web.py [hunk 18]
```diff
@@ -2856,17 +2882,17 @@ def _got_json(raw):
...
+        d.addCallback(_check_upload, "mdmf", b"URI:MDMF")
+        d.addCallback(_check_upload, "MDMF", b"URI:MDMF")
         return d
 
     @inlineCallbacks
     def test_POST_upload_bad_format(self):
         url = self.webish_url + self.public_url + "/foo?t=upload&format=foo"
         body, headers = self.build_form(file=("foo.txt", self.NEWFILE_CONTENTS * 300000))
-        yield self.assertHTTPError(url, 400, "Unknown format: foo",
+        yield self.assertHTTPError(url, 400, "Unknown format:",
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:2895]: another "t=foo"

File: src/allmydata/test/web/test_web.py [hunk 19]
```diff
@@ -3356,7 +3388,7 @@ def test_POST_mkdir_sdmf(self):
     def test_POST_mkdir_bad_format(self):
         url = (self.webish_url + self.public_url +
                "/foo?t=mkdir&name=newdir&format=foo")
-        yield self.assertHTTPError(url, 400, "Unknown format: foo",
+        yield self.assertHTTPError(url, 400, "Unknown format:",
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:3391]: "t=foo"

File: src/allmydata/test/web/test_web.py [hunk 20]
```diff
@@ -3408,8 +3440,8 @@ def test_POST_mkdir_initial_children_bad_format(self):
         (newkids, caps) = self._create_initial_children()
         url = (self.webish_url + self.public_url +
                "/foo?t=mkdir-with-children&name=newdir&format=foo")
-        yield self.assertHTTPError(url, 400, "Unknown format: foo",
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:3411]: "t=foo"

File: src/allmydata/test/web/test_web.py [hunk 21]
```diff
@@ -3487,7 +3519,7 @@ def _after_mkdir(res):
     @inlineCallbacks
     def test_POST_mkdir_no_parentdir_noredirect_bad_format(self):
         url = self.webish_url + self.public_url + "/uri?t=mkdir&format=foo"
-        yield self.assertHTTPError(url, 400, "Unknown format: foo",
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:3490]: "t=foo"

File: src/allmydata/test/web/test_web.py [hunk 22]
```diff
@@ -4428,7 +4462,7 @@ def _got_json(raw):
     def test_PUT_NEWFILEURL_bad_format(self):
         new_contents = self.NEWFILE_CONTENTS * 300000
         url = self.webish_url + self.public_url + "/foo/foo.txt?format=foo"
-        yield self.assertHTTPError(url, 400, "Unknown format: foo",
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:4431]: "t=foo"

File: src/allmydata/test/web/test_web.py [hunk 23]
```diff
@@ -4561,8 +4595,8 @@ def _got(res):
     @inlineCallbacks
     def test_PUT_mkdir_bad_format(self):
         url = self.webish_url + "/uri?t=mkdir&format=foo"
-        yield self.assertHTTPError(url, 400, "Unknown format: foo",
-                                   method="put", data="")
```
Comment (exarkun) [inline src/allmydata/test/web/test_web.py:4565]: "t=foo"

**Review by exarkun — APPROVED** [review]
Thanks.  Great progress.  Various comments inline.  I don't think there's anything particularly major.  I think the biggest deal is few changes that lessen test coverage of one particular codepath.  It occurs to me that the particular behavior may be covered elsewhere.  If that's the case then great.  If not, please bring back that coverage somehow.  Then merge.  Thanks again.

File: src/allmydata/util/jsonbytes.py [hunk 24]
```diff
@@ -13,34 +13,41 @@
...
 
+def _bytes_to_unicode(obj):
+    """Convert any bytes objects to unicode, recursively."""
+    if isinstance(obj, bytes):
+        return obj.decode("utf-8")
+    if isinstance(obj, dict):
+        new_obj = {}
+        for k, v in obj.items():
+            if isinstance(k, bytes):
+                k = k.decode("utf-8")
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/util/jsonbytes.py:27]: JSON only allows string keys, so that just errors out (though turns out Python's json support does an awful thing where it stringifies ints/floats/etc.. Not tuples or lists though).

File: src/allmydata/web/operations.py [hunk 25]
```diff
@@ -151,7 +152,7 @@ class ReloadMixin(object):
     @renderer
     def refresh(self, req, tag):
         if self.monitor.is_finished():
-            return ""
+            return b""
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/web/operations.py:155]: Hm. That might be an erroneous change. I'll try and see if disabling that is OK.

[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/web/operations.py:155]: Yeah seems like that was unnecessary. Will undo.

File: src/allmydata/web/directory.py (hunk 7 above)
Comment (itamarst) [reply src/allmydata/web/directory.py:1301]: Probably did this before the other, yeah.

File: src/allmydata/test/web/test_web.py (hunk 8 above)
Comment (itamarst) [reply src/allmydata/test/web/test_web.py:514]: Yeah, futurize throws them in and I don't always remove them.

File: src/allmydata/test/web/test_web.py (hunk 9 above)
Comment (itamarst) [reply src/allmydata/test/web/test_web.py:1764]: Yeah, it turns into `b"foo"` and I ran out of spoons to fix all those. Which is maybe a bad reason not to do it. So I guess I'll go undo this and other similar changes and fix code.

File: src/allmydata/test/web/test_web.py (hunk 11 above)
Comment (itamarst) [reply src/allmydata/test/web/test_web.py:1951]: I'm going to guess this is not actionable.

File: src/allmydata/test/web/test_web.py (hunk 12 above)
Comment (itamarst) [reply src/allmydata/test/web/test_web.py:2232]: The test was previously broken, asserting something like assertTrue(str(json["size"])) and actually doing str of a dict...

File: src/allmydata/webish.py [hunk 26]
```diff
@@ -44,6 +44,43 @@
...
+                # We use the file field to upload files, see directory.py's
+                # _POST_upload. Lack of _mime_filename means we need to trick
+                # FieldStorage into thinking there is a filename so it'll
+                # return bytes.
+                return "unknown-filename"
+            return self._mime_filename
+
+        @filename.setter
+        def filename(self, value):
+            self._mime_filename = value
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/webish.py:81]: Just going to live with this for now.
